### DISCUSSION TITLE
Selectable Index Patterns for Elasticsearch DataSources

### DISCUSSION THREAD

**Discussion by hikerspath** [grafana/grafana#56252] [Observability]
**What would you like to be added**:
Allow elasticsearch index name selection when creating a query for a panel (similar to namespace selection on a CloudWatch datasource.

**Why is this needed**:
Presently, when setting up an elasticsearch datasource, you must enter the index name as part of the datasource configuration.  This means that even though accessing the same elasticsearch cluster, I need to manually define every single index pattern as its own datasource.

If instead there was only necessity to define the timeseries pattern format (EX `(.*)-YYYY.MM.DD`), it would be very easy to provide the user with a selectable list of index patterns that match per query using the `GET /_aliases` endpoint on an elasticsearch cluster, or perhaps one of the `/_cat/` endpoints depending on how the information is parsed.

**Comment by henrikno** [comment]
Came here to report a similar thing. We have 50+ index patterns, and multiple clusters, so having to make one datasource for each means hundreds of datasources (we can script that part, but managing them in the UI is not easy).

Currently, especially with Index Lifecycle Management or aliases you don't even need to specify `-YYYY.MM.DD`. Elasticsearch is pretty good at skipping querying indices that don't match the time period.

**Comment by danielmotaleite** [comment]
Same here!
i worked around by creating a logstash-* source, but this also mean that queries are slow, because they will search in all of indexes instead of a very restricted set

**Comment by hikerspath** [comment]
I would do that, but we have almost 12000 indices and more than 40k shards housing more than 200TB of logs, so that is not realistic.

**Comment by hikerspath** [comment]
![](https://media.giphy.com/media/iDJuQR0UmiqOI/giphy.gif)

**Comment by simianhacker** [comment]
After looking the UI and some discussion with our internal users, I think there are two viable options for overriding index pattern: 

![image](https://user-images.githubusercontent.com/41702/100285903-f330e100-2f2e-11eb-870a-f4587ddf2988.png)

### Option One

For this option, we would add the index pattern to the query options. This would allow the user to override the index pattern for the entire panel, similar to how the `Min interval` works. This seems like a natural place since there are already query options overriding the datasource options. 

### Option Two (prefered)

For this options we would add an Index Pattern field next the the `Alias` pattern. This would allow you to override the index pattern per query and have the added benefit of supporting multiple index patterns on the same visualization. The other benefit of adding this to the query builder is that it will also work in the Explore interface where option one will only be applied to the panel visualizations.

With both of these options I think just having a way to define a wildcard pattern would be sufficient. The index pattern definition with the date pattern isn't really necessary in Elasticsearch 5+ due to optimizations made in that release (and beyond).

My initial thought was to add multiple patterns to the data source with a way to specify the default. This would require a lot more work because we would need to build an index pattern editor within the datasource and then have a way to select different patterns throughout Grafana. Overriding the pattern in the query builder seems like a good compromise.

@Elfo404  Thoughts?

**Comment by simianhacker** [comment]
For those following along, I started looking at the code today to see what it would take to make this happen. To my surprise it's pretty minimal, you can check out the draft PR here: https://github.com/grafana/grafana/pull/29477

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #83012. Feel free to continue the discussion around this there. Thank you!

**Comment by cjabrantes** [comment]
Hi to all the people in this issue, as this feature got stall, i believe we should make a comment in the new discussion:
https://github.com/grafana/grafana/issues/83012

to make some traction on the topic and so to help prioritise this feature.

Thanks,
